Задание выполнено

VitalyTikhonov opened the pull request.

VitalyTikhonov committed:
S13 C03 Solved the linebreak problem; started the controller distinguishing process

VitalyTikhonov committed:
S13 C04 Made the Post processing

VitalyTikhonov committed:
S13 C05 Made the users and users/:id get processing

VitalyTikhonov committed:
S13 C06 Fixed the body-parser connection issue; re-enabled user creation; encountered a card creation problem

VitalyTikhonov committed:
S13 C07 Made card creation and all cards fetch

VitalyTikhonov committed:
S13 C08 Made user profile update patch

VitalyTikhonov committed:
S13 C09 Made user avatar update patch

VitalyTikhonov committed:
S13 C10 Made liking and unliking

VitalyTikhonov committed:
S13 C11 Made link validation

VitalyTikhonov committed:
S13 C12 Reworked the url regex to match VK example URLs

VitalyTikhonov committed:
S13 C13 Switched from regex link validation the the validator plugin; started reworking error handling for non-existetnt users & cards

VitalyTikhonov committed:
S13 C14 In the very complicated process of error handling refactoring

VitalyTikhonov committed:
S13 C15 In the outer space of borderless refactoring of the error handling logic

VitalyTikhonov committed:
S13 C16 Started the crazy config attempt; revert to here if cancelled

VitalyTikhonov committed:
S13 C17 The config attempt: non-working new Function variant

VitalyTikhonov committed:
S13 C18 The config attempt: encountered the send undefined error; wrote a description to this development line

VitalyTikhonov committed:
S13 C20 Revived the idea of the CONFIG: fixed
the argument names issue, debugged the first
paths.

VitalyTikhonov committed:
S13 C21 Minor preparations before simplifying the Config model

VitalyTikhonov committed:
S13 C22 Prepared to do without the Config
To keep the Config for reference only.

VitalyTikhonov committed:
S13 C23 Making natural error handling
Re-implemented the first user controllers with orFail, managed to write a working custom orFail callback, discoveredd how to do without one.

VitalyTikhonov committed:
S13 C24 Made individual user controller handlers
and seemingly grasped the zen of orFail and error objects...

VitalyTikhonov committed:
S13 C25 A handler per controller; the User.exists
Made the User.exists check work and paused before deciding on its architecture.

VitalyTikhonov committed:
S13 C26 All controller handlers ready
About to consolidate them

VitalyTikhonov committed:
S13 C27 Consolidated controller handlers ready
Old handlers to be deleted

VitalyTikhonov committed:
S13 C28 Consolidated handlers; ObjectId checks

VitalyTikhonov committed:
S13 C29 Before final testing
Consolidated handlers complicated with additional id checks in try...catch.

VitalyTikhonov committed:
S13 C30 Edited README

zlocate commented:
Добрый день. Благодарю за проделанную работу.
К сожалению, текущая версия содержит ряд недостатков не позволяющих мне засчитать работу на данной итерации:
## Отлично
- Большой набор вдумчивых и правильных комментариев-вопросов. По мере проверки постараюсь ответить на бОльшую часть из них :)

В целом к унификации и упрощению вы должный пройти к 15му спринту, так что вы несколько опережаете время.

Де-факто исходя из вашей идеи вам следует реализовать общий .catch-блок который навешивать при помощи обертки на переданную функцию

Теперь по коду который размещаен у вас в helpers - указанные вами ошибки могут быть указаны в качестве второго параметра валидации на уровне схемы Mongoose тогда у вас отпадет необходимость итерироваться по объекту с ошибками и соотносить ошибку определенного поля со словарем - достаточно будет для ошибок валидации возвращать просто err.message, соответственно также отпадает необходимость в тяжеловестном joinErrorMessages
Гл убого его не анализировал (ибо в целом можно его исключить см выше), но на первый взгляд можно упростить его флоу если использовать метод filter :)

isUserExistent - с точки зрения логики может быть оформлена в виде мидлвары, тогда код самого контроллера станет легче. 
isObjectIdValid - также может быть реализован в виде мидлвары либо при помощи вынесения обработчика CastError в .catch-блок. Тем самым мы делаем контроллер чище.


createDocHandler - можно обобщить и сделать функцию обертку что-то вроде catchAsync которая возвращает переданную ей функцию с навешенным обработчиком в catch - как сделать догадаетесь. В принципе это чуть более абсртрактный вариант createDocHandler

createDocHandler - заменяется catchAsync'ом если вы реализуете адекватный универсальный catch-хэндлер (а в нашем api тип ошибок и реакции на них достаточно типичен и может быть переиспользован)

getLikeDeleteHandler - то же самое. Отпадет необходимость если в универсальном catch-хэндлере вы будете проверять исключение DocumentNotFoundError а кейс отсутствия документа по вашему запросу при помощи orFail (используйте его без параметров, пока не дойдете до кастомных ошибок так будет проще)
http://thecodebarbarian.com/whats-new-in-mongoose-53-orfail-and-global-toobject.html
По поводу того что он не работает - findOneAndUpdate не могу сказать на память, но емнип он работает с другими методами update-группы, которые можно использовать вместо него

## Можно лучше
- Правило `eslint`  приводит к множественным срабатываниям на Linux системах. На мой взгляд принудительно требовать соответствия переносов строки определенному системному стилю несколько избыточно (особенно учитывая что члены команды могут работать на разных системах с разными переносами.
` "linebreak-style": ["error", "windows"],`
Переносы строки можно зафиксировать например на уровне конфигурации git при помощи autoCRLF в `.gitattributes` для избежания подобных проблем
- Указать описание проекта в поле description файла package.json
- Указать автора работы (т.е вас) в поле author файла package.json.
- Пакеты `eslint-config-airbnb-base`, `eslint-plugin-import`, `nodemon` относятся к devDependencies поскольку их наличие не требуется для непосредственного запуска проекта в production.
- Старайтесь избегать закомментированного кода в финальной версии проекта
- Корректным считается запуск сервиса после подключения основных мидлвар. Сейчас же они "навешиваются" после того как мы вызовем listen
- В случае когда ваш код состоит из последовательности асинхронных вызовов в ряде случаев удобно использовать синтаксис async\await позволяющий писать асинхронный код в псевдосинхронном стиле.
- Использование "захардкоженных" статус-кодов в блоке .catch является плохой практикой поскольку может маскировать реальную проблему (код определяется вне зависимости от типа выброшенной ошибки), а также делает в ряде случаев статус-коды нерелеватными.
Для решения данной проблемы рекомендую проверять тип ошибки при помощи поля name или с использованием конструкции instanceof

- В случае если в метод удаления карточки \ простановки или снятия лайка \ запроса пользователя по ID передан некорректный по форме ID сервис возвращает не слишком информативную ошибку со статус-кодом 500, вместо более подходящей 400й (фактически это ошибка валидации).
Для этого в .catch-блоке достаточно обрабатывать ошибку с именем CastError.


В целом действительно решение получилось несколько перегруженным, но рабочим. Поэтому я принимаю работу. 
Если у вас возникнут вопросы по моим комментариям - не стесняйтесь задавать их в PR'е постараюсь помочь, хоть и не гарантирую оперативность. (множество работ на проверке)

VitalyTikhonov committed:
Saving files before refreshing line endings

VitalyTikhonov committed:
Line-endings re-configuration Attempt 3
Added .gitattributes (text=auto, *.* text eol=lf, binaries specified).
Reverted .editorconfig to end_of_line = lf.
Disabled the "linebreak-style": ["error", "windows"] rule in .eslintrc.js.
git add --renormalize . seemed to have not worked: still linting errors after commit and push.
Fixed them with eslint ./ --fix.

VitalyTikhonov committed:
Removed .gitattributes from .gitignore

VitalyTikhonov committed:
Edited package.json, reinstalled dependencies
Added description and authorship to package.json.
Reinstalled eslint-config-airbnb-base, eslint-plugin-import and nodemon as dev dependencies.

VitalyTikhonov commented:
Добрый день! Благодарю за развернутый комментарий.
Часть рекомендаций реализовал, другую часть, связанную с рефакторингом, проработаю позже, на этапе подготовки к следующему проекту, так не успел сделать это до начала спринта.
